add migration CLI command to update viewconfs

sergpolly opened the pull request:
## Description
when migrating an existing higlass-instance to a new location (new URL)  - everything works except the `viewconfs` saved in the database, because they refer to tileset via old URLs ...

So, we gotta replace `origin` to `destination` in those `viewconfs`, i.e. in the database

What I mean by "migration":
simply copying content of `hg-data` and restarting docker on a new machine/URL with `--data-dir` pointing to that copied `hg-data`.

#### What was changed in this pull request?

 - added a `migration` CLI command
 - added `get_site_url` function to `.common` module
 - implemented a first working prototype of `migration.py`


#### Why is it necessary?

some people maintaining their own higlass-es sometimes move them
e.g. we plan to move ours from aws to local machine open to the world - `SITE_URL` is going to change - but we want `viewconfs` to keep working, e.g.: `http://higlass.dekkerlab.org/app/?config=Yk_ArP04RlWiGtReKkPAbA` would turn into `http://newserver.edu:12345/app/?config=Yk_ArP04RlWiGtReKkPAbA` and keep working ...

Fixes #53 

## Checklist

- [x] Unit tests added or updated
- [x] Documentation added or updated
- [ ] Updated CHANGELOG.md
- [x] Run `black .`

sergpolly commented on higlass_manage/migrate.py:
```diff
@@ -0,0 +1,190 @@
+    # sanitize it to have a http://blah.blah.domain:XXXX form
+
+    # # use something like that to validate URLs ...
+    # pieces = urlparse.urlparse(url)
```
what's the best/non-invasive way to sanitize a URL in this context?

pkerpedjiev replied:
Why do you need to sanitize the URL?

Also, have you considered including the schema (http, https)?

sergpolly replied:
maybe "sanitize" is a bad word - what I meant - is simply to make sure that `origin` and `destination` - after we generate them have a form `[schema]://whatever:[port]`

that `higlass_manage.common.get_site_url` returns URL with the `schema` already

pkerpedjiev replied:
Format seems fine to me. You could also use f-strings: `f"{server}:{port}"`

sergpolly commented on higlass_manage/migrate.py:
```diff
@@ -0,0 +1,190 @@
+    # there is a nice built-ion backup function in Python >=3.7:
+    # this is just stupid mock code - cause i don't know
+    # how get around Python version issue nicely ...
+    _python_version = 3.6
```
this is just super-ugly right now , because it's unclear what should be the way for `Python<3.7`
also how to write Python-version dependent  blocks like that

pkerpedjiev replied:
I would personally omit the version check and use the "ugly" python 3.6 version.

This isn't a crazy high-availability server. Stopping the running instance and copying the db is perfectly fine, IMO.

sergpolly replied:
understood will do, - it's just `db_conn1.backup(db_conn2)` looks sooo much better, and it can be done "alive" - while the database is in use (as far as I understood) 

i'll leave a comment about `db_conn1.backup(db_conn2)` for future references

sergpolly commented on higlass_manage/migrate.py:
```diff
@@ -0,0 +1,190 @@
+    except Error as e:
+        print(e)
+
+    sql = '''UPDATE tilesets_viewconf
```
this is the essence of what's going on here - all we want to do is to replace `origin` with `destination` in the `viewconf` table of the database...

pkerpedjiev commented on higlass_manage/migrate.py:
```diff
@@ -0,0 +1,190 @@
+        con.close()
+    else:
+        # OMG - this is ugly ...
+        def _stop(name):
```
We have a stop function here: https://github.com/higlass/higlass-manage/blob/687b40f56fca3c392f401b294828a8c0f9f66c5c/higlass_manage/stop.py#L10

You can just factor that out and use it here as well.

sergpolly replied:
I did that first, but ...
`higlass_manage.stop.stop` - stops and removes the higlass-container itself, then it tries to kill redis and the associated docker-nework - redis and network throw exceptions and ... basicaly what I did was - just something to make it work

Should I still reuse `higlass_manage.stop.stop` but catch redis/network exceptions ?
also removing container is ... not cool , maybe - although if someone is going to migrate it anyways - maybe it's fine

pkerpedjiev commented:
Looks like a nice first pass. Throw some tests and a README.md update with an example of how to run it and it'll be good shape.

sergpolly commented on higlass_manage/migrate.py:
```diff
@@ -0,0 +1,190 @@
+    # backup the sqlite3 database:
+    # a better way to do a backup ...
+    # https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.backup
+    data_dir = get_data_dir(hg_name)
```
right now  this is the only place that requires "alive" higlass-instance (running or `Exited`) ...
Since we've provided `--origin-site` and `--origin-port` options - I guess we should also add `--data-dir`, such that the "migration" could be performed on a "dead body" ...

pkerpedjiev commented on README.md:
````diff
@@ -138,6 +138,27 @@ To stop a running instance, use the `stop` command:
+```bash
+# at the destination:
+scp -r user@origin:/path/to/hg-data  /new/path/
+higlass-manage start --data-dit /new/path/hg-data ...
````
Small typo: should be `--data-dir`.

pkerpedjiev commented:
One more other idea upon reading your README. Would it make sense to have the procedure be something like this?

1. Copy over data dir
2. Run a migrate command that looks like this `higlass-manage migrate --new-data-dir /new/data/dir --old-site-url old.host.org --new-site-url new.host.org`?

It might make it more explicit where all the information is coming from? Perhaps even better might be to rename the `migrate` command to something more explicit like `update-viewconfs` so that it's clear to the user what they're doing.

`higlass-manage migrate --new-data-dir /new/data/dir --old-site-url old.host.org --new-site-url new.host.org`

sergpolly commented:
> It might make it more explicit where all the information is coming from? Perhaps even better might be to rename the `migrate` command to something more explicit like `update-viewconfs` so that it's clear to the user what they're doing.

1. yep - agreed! migrate -> update-viewconfs - I did `update_viewconfs` not sure where to enforce `update-viewconfs` ...
2. I propose to add symmetry to the CLI api - layed out in README - either use a running higlass to infer site-url, port and data-dir for update OR provide those explicitly - could be done on both "old" and "new" hosts [not implemented yet]
3. `db.sqlite3` should have viewconfs updates, `db.sqlite3.backup` should be the original copy [not implemented yet] ...

if the concept is ok I'll hopefully finish that in a day or two

pkerpedjiev commented:
> 1. yep - agreed! migrate -> update-viewconfs - I did update_viewconfs not sure where to enforce update-viewconfs ...

`update_viewconfs` is fine.

> 2. I propose to add symmetry to the CLI api - layed out in README - either use a running higlass to infer site-url, port and data-dir for update OR provide those explicitly - could be done on both "old" and "new" hosts [not implemented yet]

I'm not sure I completely understand what you mean by add symmetry. I personally think the easiest and clearest solution is to just provide the most explicit version:

`higlass-manage update-viewconfs --old-site-url http://old.host.org --data-dir /old/path/to/data --new-site-url http://new.host.org`

This way there's no uncertainty about what's being changed and the existing instance is unaffected so the new one can be tested while the old one is running.

sergpolly commented:
i just thought it's kind of convenient to be able to do both before you copy stuff (on the `old.host`) or even after you copy the stuff, i.e. on the `new.host` - and also depending on the running-status of the higlass instance one could have inferred site-url,port either from the running instance or by manually providing it ...

this is getting too complidcated, i guess - i'll the version you proposed and then we'll see

sergpolly commented:
cannot reuse `higlass_manage.stop.stop()` as Python API, because it was defined as `@click.command()` and those cannot be reused as Python API (ever, I guess https://github.com/pallets/click/issues/40)

pkerpedjiev commented:
That's pretty easy to get around by using the `click` function as a simple wrapper for the actual function. See here: https://github.com/higlass/higlass-manage/blob/687b40f56fca3c392f401b294828a8c0f9f66c5c/higlass_manage/start.py#L134

sergpolly commented:
yes ... I was just trying to minimize amount of changes I make to the rest of `higlass-manage` ...
also I suspected that I couldn't reuse `click.command` as python API but didn't know it for sure

sergpolly commented:
slowly getting there ...

sergpolly commented:
@pkerpedjiev could you help me with tests ?
ideally we'd save some viewconf as a link `Export View as Link`, get `uid` of that link, then run `update-viewconfs` and check the `dbsqlite3.updated` ..., right ?

Should I bother with "simulating"  `Export View as Link` behavior or could I just ingest something into `db.sqlite3` and then try to modify it ?!

sergpolly commented:
ha, looks like I can make an entry in the `db.sqlite3` by posting
```
requests.post("http://localhost:12345/api/v1/viewconfs/",json={"viewconf":SOME_CONFIG_WE_WANT_TO_SAVE})
```
...

sergpolly commented:
minimal working json - stollen from https://github.com/higlass/higlass-python/blob/master/notebooks/Examples.ipynb - this viewconf could be used in conjunction with this https://github.com/higlass/higlass-manage/blob/687b40f56fca3c392f401b294828a8c0f9f66c5c/test.sh#L44 ...
```json
{
  "editable": true,
  "views": [
    {
      "uid": "myviewconf123",
      "tracks": {
        "top": [],
        "center": [
          {
            "type": "heatmap",
            "options": {
              "valueScaleMax": 0.5
            },
            "tilesetUid": "a",
            "server": "http://localhost:8123/api/v1/",
            "height": 250
          }
        ],
        "left": [],
        "right": [],
        "bottom": []
      },
      "layout": {
        "w": 12,
        "h": 6,
        "x": 0,
        "y": 0
      }
    }
  ],
  "trackSourceServers": [
    "http://localhost:8123/api/v1/"
  ],
  "locationLocks": {
    "locksByViewUid": {},
    "locksDict": {}
  },
  "zoomLocks": {
    "locksByViewUid": {},
    "locksDict": {}
  },
  "exportViewUrl": "http://localhost:8123/api/v1/viewconfs"
}
```

sergpolly commented:
@pkerpedjiev it worked !
my `viewconf`  POSTing experiment worked out and `update-viewconfs` testing is in `test.sh` now!
the only thing is that I had to moodify `get_test_data.sh` to add my `test_viewconf.json` there - it is still a mystery to me - how do you get  `data/default.json` there?! - I'd do it the same way if I knew how: https://github.com/dekkerlab/higlass-manage/blob/fb80453f9de54a2ebd250ab3cfa740d032f4e299/test.sh#L52

I think this is in a pretty good shape! what do you think ?

sergpolly commented:
ohh - `data/default_options.json` is already in the `data` folder - it's just the `data` folder is not tracked/ignored by `git`, because  of `.gitignore`: https://github.com/higlass/higlass-manage/blob/687b40f56fca3c392f401b294828a8c0f9f66c5c/.gitignore#L8

is it intentional or can we overrule this and put the `test_viewconf.json` in the `data` directly ?

sergpolly commented on higlass_manage/update_viewconfs.py:
```diff
@@ -0,0 +1,183 @@
+            )
+        sys.stderr.flush()
+        # not so safe way to backup a database:
+        shutil.copyfile(origin_db_path, update_db_path)
```
should we get rid of `shutil.copyfile` altogether ?!:
 - we rely on `sqlite3` anyhow, right ?
 - `sqlite3` python module implies `sqlite3` executable, - not sure ?
 - if I understand it correctly `es = sp.run(["sqlite3",origin_db_path,f".backup {update_db_path}"])` - should be "safest", it could be even done, while higlass-container is running - right ?

pkerpedjiev replied:
I agree that removing one of the methods is better. It'll simplify the implementation. The most likely reason one of these would fail anyway is because a disc runs out of space. And then both would fail.

If it were up to me though, I'd copy the sqlite3 file rather than using sqlite3's backup function. My reasoning is just that it's a simpler operation that's guaranteed to be supported on all systems.

I won't insist though. If you'd prefer to use backup, that's fine too.

sergpolly commented on higlass_manage/update_viewconfs.py:
```diff
@@ -111,48 +102,45 @@ def update_viewconfs(old_hg_name,
-                    "The database backup using sqlite3 exited"
-                    f"with error code {res.returncode}."
-                )
+                "The database backup using sqlite3 exited"
```
so, right now it's like so:
```python
try:
    # backup db.sqlite3 using sqlite3 CLI
except OSError:
    # presumably sqlite3 was not found
    # shutil copyfile - not safe anyways ...
else:
    # no exceptions
    # check return status of sqlite3 .backup
    if returncode != 0:
        # fail HARD, - because WTF ?
```
so, it matches :
> if the sqlite3 backup command fails we just copy the DB.
or am I missing something ?

sergpolly replied:
besides -  I think we shouldn't be using `shutil.copyfile` , cause we rely on `sqlite3` anyhow (at least via `clodius`)

What do you think ?

pkerpedjiev replied:
The reason I mentioned that is because you're checking the return code of the backup command outside of the try/except block. Or maybe I'm mistaken, the indentation is a little hard for me to see in the diffs.

sergpolly commented on higlass_manage/update_viewconfs.py:
```diff
@@ -0,0 +1,171 @@
+    # once update_db_path is backedup, we can connect to it:
+    conn = None
+    try:
+        conn = sqlite3.connect(update_db_path)
```
I'm modifying `update_db_path`, - not the original one

pkerpedjiev replied:
Ah, right. For some reason I had it in my head that you were creating db.sqlite3.backup and updating the original. My bad.

sergpolly commented on higlass_manage/stop.py:
```diff
@@ -7,9 +7,44 @@
+    )
+
+
+def _stop(
```
added some granularity to `stop`, such that the container could be "paused" - by doing `container.stop()` instead of `container.stop(); container.remove()`

sergpolly commented:
> You stop a running higlass server to update it. Would it make sense to start it again after completing the operation? It looks like you have a comment mentioning that .

yes, I think we should restart - I just wasn't sure which way is the best - I wanted to avoid explicit `docker` calls in the script - want to reuse things as much as possible  - see comments below

> How do you feel about adding an option where the user can specify the name of the backup db.  Or perhaps even having it the other way around where the original DB remains unchanged and the newly created DB has the new server in the viewconfs?

that's exactly how it is now - we backup the original DB named `db.sqlite3` to `db.sqlite3.updated`, and then modify `db.sqlite3.updated` - I added a comment there https://github.com/higlass/higlass-manage/pull/54/files#r368201674

> If you do this, maybe you could pause the container instead of stopping it and then just restart as before?

that was my intent as well - I do only "pause" container - I modified `stop` for that - see comment: https://github.com/higlass/higlass-manage/pull/54/files#r368201903

We just don't have an existing `restart` functionality - which could be added to the `_start` itself, right ? - something like that: add option `hg-name` to `start` - then check if `hg-name` already exist but not running - than restart it ; or name the option `--restart-hg-name` ...

> How do you feel about adding an option where the user can specify the name of the backup db.

not a problem - will add, once we resolve conceptual question

pkerpedjiev commented:
>yes, I think we should restart - I just wasn't sure which way is the best - I wanted to avoid explicit docker calls in the script - want to reuse things as much as possible - see comments below

Nothing wrong with explicit Docker calls in the script. There's a pause function here: https://docker-py.readthedocs.io/en/stable/containers.html#docker.models.containers.Container.pause

Should just be a matter of getting the container by name and then pausing and restarting it?

https://github.com/higlass/higlass-manage/blob/687b40f56fca3c392f401b294828a8c0f9f66c5c/higlass_manage/stop.py#L23

So, to summarize:

> shutil.copyfile or not

I vote for shutil.copy instead of backup :-) If not, then use backup and no copyfile.

> how best to perform restart

I vote for `container.pause()` and then `container.unpause()`. Seems like the simplest solution. No need to worry about starting with all the correct parameters and no chance of the DB being modified while it's paused.

> not 100% sure, but I guess sqlite3 DB .backup should be "safe" even when DB is in use

> in the light of (3) - maybe there is no need to stop,restart the container ?

If that's the case, then so much easier!

sergpolly commented on higlass_manage/update_viewconfs.py:
```diff
@@ -115,38 +114,37 @@ def update_viewconfs(
+        if old_hg_name is not None:
+            sys.stderr.write(f"Restarting container {old_hg_name} ...\n")
+            sys.stderr.flush()
+            _name = "{}-{}".format(CONTAINER_PREFIX, old_hg_name)
```
this of course looks a bit ugly - we might consider baking `restart` capabilities in `higlass-manage start` or providing separate  `higlass-manage restart` CLI - but that's probably for later

sergpolly commented:
should be more or less final now

tests are failiing because `pysam` (requirement of `clodius`) cannot find `cython` or some `.c` file - something got updated on `travis`-side ?

sergpolly commented:
clodius version bump - did not help - it's unclear why clodius tests (and pysam installation) working just fine...

But I checked on PyPI - indeed pysam versions before 0.15.4 had `pysam/libc*.c` files pre-packaged in their `tar.gz` source archives - but not with `0.15.4` - why does clodius-travis and higlass-manage-travis install pysam differently - idk

But I just wanted to test it - and did some mess there https://github.com/higlass/clodius/pull/121

sergpolly commented:
adding `pysam` as an explicit requirement - fails as well - maybe it has something to do with the travis differences between clodius and higlass-manage - which there are plenty

I cannot reproduce this on my local machine though

pkerpedjiev commented:
Can you try using this `.travis.yml`? https://github.com/higlass/higlass-manage/blob/pkerpedjiev/better-parameter-documentaiton/.travis.yml

I updated it for another PR and it seems to work.

sergpolly commented:
all set ! should be good to go now ...

pkerpedjiev commented on higlass_manage/stop.py:
```diff
@@ -7,9 +7,44 @@
+    "--remove-network-bridge/--dont-remove-network-bridge",
+    default=True,
+    show_default=True,
+    help="Remove netowrk bridge associated with a given higlass instance.",
```
```suggestion
    help="Remove network bridge associated with a given higlass instance.",
```

pkerpedjiev commented on higlass_manage/update_viewconfs.py:
```diff
@@ -0,0 +1,194 @@
+    "--old-site-url",
+    help="site-url at the old location."
+    " Provide this when higlass container"
+    " one is updating is not running.",
```
This phrasing is a little easier for me to parse:

```suggestion
    " to be updated is not running.",
```

pkerpedjiev commented on higlass_manage/update_viewconfs.py:
```diff
@@ -0,0 +1,194 @@
+    "--old-port",
+    help="port at the old location."
+    " Provide this when higlass container"
+    " one is updating is not running.",
```
Same as previous.

```suggestion
    " to be updated is not running.",
```

pkerpedjiev commented on higlass_manage/update_viewconfs.py:
```diff
@@ -0,0 +1,194 @@
+    "--db-backup-name",
+    help="name of the database (db) backup file."
+    " db backup will be stored in the data directory"
+    " provided explicitly or inferred from running"
```
```suggestion
    " provided explicitly or inferred from the running"
```

pkerpedjiev commented on higlass_manage/update_viewconfs.py:
```diff
@@ -0,0 +1,194 @@
+    # were updated
+
+    sys.stderr.write(
+        f"Backedup version of the database {update_db_path}\n"
```
```suggestion
        f"Backed up version of the database {update_db_path}\n"
```
